sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -315,4 +315,65 @@ case class StringLength(child: Expression) extends UnaryExpression with ExpectsI
   }
 }
 
+/**
+ * Like ConcatWS below, but taking an array of strings.
+ */
+case class ConcatWS(sep: Expression, child: Expression)
+  extends Expression with ExpectsInputTypes {
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def foldable: Boolean = child.foldable && sep.foldable
+  override def children: Seq[Expression] = sep :: child :: Nil
+  override def expectedChildTypes: Seq[DataType] = StringType :: ArrayType(StringType) :: Nil
+  override def toString: String = s"""CONCAT_WS($sep, $child)"""
 
+  override def eval(input: Row): Any = {
+    val sepEval = sep.eval(input)
+    val childArr = child.eval(input)
+    if (sepEval != null && childArr != null) {
+      val separator = sepEval.asInstanceOf[UTF8String].toString
+      val validSeq = childArr.asInstanceOf[Seq[UTF8String]]
+      if (!validSeq.contains(null)) {
+        return UTF8String.fromString(validSeq.mkString(separator))
+      }
+    }
+    null
+  }
+}
+
+/**
+ * Like Concat below, but with custom separator SEP.
+ */
+object ConcatWS {
+  def apply(sep: Expression, exprs: Expression*): ConcatWS = {
+    apply(sep, CreateArray(exprs))
+  }
+}
+
+/**
+ * A function that returns the string or bytes resulting from concatenating the strings or bytes
+ * passed in as parameters in order. For example, concat('foo', 'bar') results in 'foobar'. Note
+ * that this function can take any number of input strings.
+ */
+case class Concat(child: Expression*)
+  extends Expression with ExpectsInputTypes {
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = child.exists(_.nullable)
+  override def foldable: Boolean = child.forall(_.foldable)
+  override def children: Seq[Expression] = child
+  override def expectedChildTypes: Seq[DataType] = Seq.fill(child.size)(StringType)
+  override def toString: String = s"""CONCAT($child)"""
+
+  override def eval(input: Row): Any = {
+    val validSeq = if (child != null && !child.contains(null)) {
+      child.map(_.eval(input))
+    } else {
+      Seq()
+    }
+    if (validSeq.nonEmpty && !validSeq.contains(null)) {
+      UTF8String.fromString(validSeq.mkString(""))
```

Comment:
I tried `select concat(null, null);` and the return value is `null`;
but `concat()` with empty-parenthesis does return `""`;